src/bootstrap/src/core/builder.rs

```diff
@@ -72,36 +72,40 @@ impl<'a> Deref for Builder<'a> {
 }
 
 pub trait Step: 'static + Clone + Debug + PartialEq + Eq + Hash {
-    /// `PathBuf` when directories are created or to return a `Compiler` once
-    /// it's been assembled.
+    /// Result type of `Step::run`.
     type Output: Clone;
```

Comment:
Discussion: ah I happen to know a bit about how `Output` associated type is used...

- Very often, if the `Step` is not intended to provide any kind of "wiring up" info to dependent `Step`s, this will just be `()`.
- Sometimes a `Step` will provide a `PathBuf` for various reasons, often the path to the root directory of the built artifacts (or path of a built artifact itself).
- Sometimes like `compiler::Rustc`, it will be `u32` as it is intended to indicate a compiler stage (... subject to off-by-one trickery).

I wonder if we should describe examples here, or if that will make it more confusing...

Reply:
`Result type of Step::run` should be dead simple already.